File: 백준/Silver/15723. n단 논법/n단 논법.js
const fs = require('fs');
let input = fs.readFileSync('/dev/stdin').toString().trim().split('\n');
let alphabet = 'abcdefghijklmnopqrstuvwxyz';
const INF = parseInt(1e9);    

/** Input 부 */
const n = parseInt(input[0]);
const n_info = input.slice(1,n+1);
const m = parseInt(input[n+1])
const m_info = input.slice(n+2);

let graph = Array.from(Array(26),() => Array(26).fill(INF));
// n개 관계 입력
n_info.forEach((tmp) => {
    let [s,e] = tmp.trim().split(" is ");
    let a = alphabet.indexOf(s);
    let b = alphabet.indexOf(e);
    graph[a][b] = 1;
});

// floyd-warshall
for(let k=0; k<26; k++){
    for(let i=0; i<26; i++){
        for(let j=0; j<26; j++){
            graph[i][j] = Math.min(graph[i][j],graph[i][k]+graph[k][j]);
        }
    }
}

m_info.forEach((tmp) => {
    let [r,c] = tmp.trim().split(" is ");
    let mx = alphabet.indexOf(r);
    let my = alphabet.indexOf(c);
    graph[mx][my] === INF ? console.log('F') : console.log('T');
});

File: 백준/Silver/15723. n단 논법/n단 논법.py
import sys
input = sys.stdin.readline
INF = int(1e9)

n = int(input().rstrip())
alphabet = 'abcdefghijklmnopqrstuvwxyz'
graph = [[INF]*26 for _ in range(26)]

# n개의 전제를 깔기
for _ in range(n):
    tmp = input().rstrip().split(" is ")
    a = alphabet.index(tmp[0])
    b = alphabet.index(tmp[1])
    graph[a][b] = 1

for k in range(26):
    for i in range(26):
        for j in range(26):
            graph[i][j] = min(graph[i][j],graph[i][k]+graph[k][j])

m = int(input().rstrip())
for _ in range(m):
    tmp = input().rstrip().split(" is ")
    a = alphabet.index(tmp[0])
    b = alphabet.index(tmp[1])

    if graph[a][b] == INF:
        print('F')
    else:
        print('T')